Playwright: restructure theme specs

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to restructure theme-related specs on the Playwright side of things.

Key changes:
- combination of the previously separate `preview` and `activate` portions of the specs into one flow.
- move the `closePreview` method to the appropriate location in `PreviewComponent`.
- new spec to replicate Selenium behavior by testing for `preview` from the theme popover in the gallery.

Details:

Previously, the two flows tested - `preview` and `activate` - were in separate specs. 
This led to a couple of things:
- wasted time and processing power (repeated login and navigation to Themes page)
- on rare occasions, conflict.

The conflict was caused by the `activate` theme selecting then activating a theme (say, Twenty Twenty-One) while the `preview` theme wanted to preview the same theme. An activated theme will not have the `Preview` button in the theme details page, being replaced with the `Customize theme` button. This led to the `preview` spec failing unexpectedly.

#### Testing instructions

TeamCity
- ensure that theme specs are not causing any failures.

worldomonation committed:
Combine the test for theme preview and theme activation into one.

worldomonation committed:
Move the closePreview method to the PreviewComponent.

worldomonation committed:
Rethink how the specs are designed for themes.
- one spec to handle preview and activation of theme from the ThemeDetails page.
- one spec to handle preview from the popover menu.

worldomonation committed:
Apply prettier to files.
Remove `viewInfo` method, instead opting to use the `clickPopoverItem` for all interactions with theme cards.

worldomonation committed:
Call PreviewComponent.closePreview() as reusable method.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -13,6 +14,10 @@ const selectors = {
+
+	// Theme card
+	popoverButton: '.theme__more-button',
+	popoverMenuItem: '.popover__menu-item',
```
New selectors for the `meatball menu` and the associated popover.

worldomonation commented on packages/calypso-e2e/src/lib/components/preview-component.ts:
```diff
@@ -20,4 +23,13 @@ export class PreviewComponent extends BaseContainer {
+	 */
+	async closePreview(): Promise< void > {
+		await this.page.click( selectors.closeButton );
+	}
```
Previously, this method was not implemented - the test ended once the theme preview was rendered.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -68,14 +73,16 @@ export class ThemesPage extends BaseContainer {
 	 *
 	 * @param {string} [name] Theme name to select.
-	 * @returns {Promise<void>} No return value.
+	 * @returns {Promise<ElementHandle>} Reference to the selected theme card on the gallery.
```
`select` method is now only for obtaining the ElementHandle to a theme matching the criteria where previously it performed additional actions (clicking into the theme to show details).

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -86,13 +93,21 @@ export class ThemesPage extends BaseContainer {
+	 * @param {string} action Action to be called from the popover.
+	 */
+	async clickPopoverItem(
+		selectedTheme: ElementHandle,
```
I've decided on having this method accept an ElementHandle as a parameter instead of calling `this.select` within the method for the following reasons:

- less ambiguity: it is clearer that this method will be called on a specific theme card. 
- decouples `select` and `action` logic: even if the `select` call is modified again in the future it should not be necessary to modify all methods that depend on `select`.

worldomonation commented on test/e2e/specs/specs-playwright/wp-theme__popover-preview.js:
```diff
@@ -28,21 +28,22 @@ describe( DataHelper.createSuiteTitle( 'Theme: Preview' ), () => {
-	it( 'Preview theme', async function () {
-		await themesDetailPage.preview();
+	it( 'Close preview', async function () {
+		await previewComponent.closePreview();
```
This file has been repurposed to test this flow:
- render themes gallery
- search for a clear theme name
- click on meatball menu
- trigger preview component from the popover

dpasque commented on packages/calypso-e2e/src/lib/components/preview-component.ts:
```diff
@@ -3,6 +3,9 @@ import { BaseContainer } from '../base-container';
 	previewPane: '.web-preview',
+
+	// Actions on pane
+	closeButton: 'button:text("Close")',
```
Definitely optional suggestion here...

Would using `button[aria-label='Close preview']` potentially be more resilient here than the text selector?

My only wondering is if in future someone adds a button to the page behind this preview that says something like "Close xyz", it would hijack this selector. 

Aria-labels carry their own risk too though... What do you think?

worldomonation replied:
Good question!

[Playwright docs](https://playwright.dev/docs/selectors/#prioritize-user-facing-attributes) recommends _user-facing attributes should be prioritized_ and includes both text and aria attributes in the example.

You raise an excellent point about another `Close xyz` button on the page behind the PreviewComponent.
It wouldn't be an issue if we prefixed the selector with the top-level CSS selector for the PreviewComponent, but that would be violating the best practice.

I'll make your suggested changes in the next push :)

dpasque commented on test/e2e/specs/specs-playwright/wp-theme__details-preview-activate.js:
```diff
@@ -3,14 +3,16 @@ import {
+	let previewComponent;
 	let popupTab;
 	let page;
 	const themeName = 'Twenty Twen';
```
To supplement the goal of keeping these tests safe to run side-by-side, what do you think of adding comments to both test files to where you have this constant of the theme name to help explaining the reasoning behind the themes used?

I.e. to explain here that this partial match is explicitly used to always have two to oscillate between, and in the other script that that twenty seventeen is used because it will never be active.

I just want to make sure no one comes along and swaps out the theme name under test without realizing the care in picking the themes under test!

worldomonation replied:
That's totally fair. I will include a comment on this.

dpasque commented on test/e2e/specs/specs-playwright/wp-theme__details-preview-activate.js:
```diff
@@ -30,22 +32,32 @@ describe( DataHelper.createSuiteTitle( 'Theme: Activate' ), () => {
-		await themesPage.select( themeName );
+	it( `Select and view details of a theme starting with ${ themeName }`, async function () {
+		const selectedTheme = await themesPage.select( themeName );
+		await themesPage.clickPopoverItem( selectedTheme, 'Info' );
```
Just as a double check: do you think we're losing anything in terms of useful coverage by using the popover here instead of clicking directly on the theme card itself? I imagine that most users use the latter more than former - do we want to reflect that in this test?

worldomonation replied:
I debated back and forth on this a bit before pushing this revision up. 

Yes, by only using the meatball menu this would indeed reduce coverage of what you and I suspect would be the more typical path for users to traverse. 

Let's restore the original method (hover then click) and use it in the `wp-theme__details-preview-activate` spec so we'd have at least one coverage.

dpasque reviewed (changes requested):
Overall looks great! I like the refactors you made to the theme page!

worldomonation committed:
Address Dan's PR review comments.
- add comment to theme name choice.
- restore the hover action in themes gallery.

dpasque reviewed (approved):
LGTM!  🚢
